File: days/day_1/task_2.py
from task_1 import parse_input


def sliding_window(list: list[int], size: int):
  return (list[i: i + size] for i in range(len(list) - size + 1))


def count_sliding_depth_increases(report: list[int]):
  count = 0
  prev_depth = -1
  for depths in sliding_window(report, 3):
    depth = sum(depths)
    if prev_depth != -1 and depth > prev_depth:
      count += 1
    prev_depth = depth
  return count


if __name__ == "__main__":
  report = parse_input('input.txt')
  count = count_sliding_depth_increases(report)
  print(count)

File: src/days/day4/task2.ts
import esMain from "es-main";
import * as R from "ramda";
import { parseInput, Pair, Range } from "./task1";

const isOverlapping = (pair: Pair<Range>) => {
  const [[s1, e1], [s2, e2]] = pair;

  return (e1 >= s2 && s1 <= e2) || (e2 >= s1 && s2 <= e1);
};

const calcOverlappingPairs = (pairs: Pair<Range>[]) => {
  return R.count(isOverlapping, pairs);
};

if (esMain(import.meta)) {
  console.log(calcOverlappingPairs(await parseInput("./test.txt")));
  console.log(calcOverlappingPairs(await parseInput("./input.txt")));
}
